garage/tf/models/base.py

```diff
@@ -0,0 +1,281 @@
+        # All TF models need a valid unique name
+        # TODO(ahtsan): produce a message on collision?
+        self._name = name or self._name
+        self._is_parent = True
+        self._is_built = False
+        # All ops and variables for this Model *MUST* be placed under this scope
+        self._variable_scope = tf.variable_scope(self.name, reuse=False)
+        self._name_scope = tf.name_scope("default")
+        self._siblings = {}
+
+    @classmethod
+    def _new_child(cls, other, name):
+        twin = cls(**other.init_spec[1])
+        twin._name = name
+        twin._is_parent = False
+        twin._default_parameters = other._default_parameters
+        twin._is_built = other._is_built
+        twin._variable_scope = other.variable_scope
+        twin._name_scope = tf.name_scope(name)
+        return twin
+
+    def build(self, *inputs, name=None):
+        if not name:
+            if 'self' in self._siblings:
+                raise ValueError("Subgraph {} already exists!".format(
+                    self.name))
+            with self._name_scope:
+                _variable_scope = self.variable_scope if not self._is_built else tf.variable_scope(
+                    self._variable_scope._name_or_scope, reuse=True)
+                with _variable_scope:
+                    self._inputs = inputs
+                    outputs = self._build(*inputs)
+                    self._outputs = list(outputs)
+                    self._siblings['self'] = self
+                    if not self._is_built:
+                        variables = self._get_variables().values()
+                        tf.get_default_session().run(
+                            tf.variables_initializer(variables))
+                        if self._default_parameters:
+                            self.parameters = self._default_parameters
+                        self._is_built = True
+        elif name in self._siblings:
+            raise ValueError("Subgraph {} already exists!".format(name))
+        elif not self._is_parent:
+            raise ValueError("Child of a model should not be built!")
+        else:
+            twin = type(self)._new_child(self, name)
+            setattr(self, name, twin)
+            self._siblings[name] = twin
+            outputs = twin.build(*inputs)
+            if not self._is_built:
+                self._inputs = inputs
+                self._outputs = list(outputs)
+                self._is_built = True
+
+        return outputs
+
+        # Initialize our variables so that they're guaranteed to exist after
+        # build
+        #
+        # NOTE: It's very important that we don't run
+        # `tf.global_variable_initializer()` after this step (which will call
+        # the mostly-random initializers on each variable, destroying the
+        # presets loaded from the file)
+        # @gautams has already hit this problem in his own code.
+        #
+        # TODO(ahtsan): We can deal with this by sanitizing
```

Comment:
Sanitization is not a real solution, because in practice people only run the sanitizer after they are done debugging. Since apparently people are already hitting this bug, I think we should do something else to fix it.

Is there any way we can dynamically detect that `tf.global_variable_initializer()` was called and throw an exception?

Reply:
We could compare the output of https://www.tensorflow.org/api_docs/python/tf/global_variables and https://www.tensorflow.org/api_docs/python/tf/report_uninitialized_variables and ensure they are 1:1.